common/pose.h

```diff
@@ -1,23 +1,209 @@
+    {}
+
+    operator Vector3<LengthUnit>() const { return Vector3<LengthUnit>(x(), y(), z()); }
+
+    LengthUnit &x() { return (*this)[0]; }
+    const LengthUnit &x() const { return (*this)[0]; }
+    LengthUnit &y() { return (*this)[1]; }
+    const LengthUnit &y() const { return (*this)[1]; }
+    static constexpr LengthUnit z() { return LengthUnit(0); }
+};
+
+//! 3D length unit vector
+template<typename LengthUnit>
+class Vector3
+  : public VectorBase<LengthUnit, 3>
+{
+public:
+    Vector3() = default;
+
+    Vector3(LengthUnit x, LengthUnit y, LengthUnit z)
+      : VectorBase<LengthUnit, 3>(x, y, z)
+    {}
+
+    Vector3(const std::array<LengthUnit, 3> &array)
+      : Vector3(array[0], array[1], array[2])
+    {}
+
+    operator Vector2<LengthUnit>() const { return Vector2<LengthUnit>(x(), y()); }
+
+    LengthUnit &x() { return (*this)[0]; }
+    const LengthUnit &x() const { return (*this)[0]; }
+    LengthUnit &y() { return (*this)[1]; }
+    const LengthUnit &y() const { return (*this)[1]; }
+    LengthUnit &z() { return (*this)[2]; }
+    const LengthUnit &z() const { return (*this)[2]; }
+};
+
+template<typename LengthUnit, typename AngleUnit>
+class Pose3;
+
+//! A two-dimensional pose
+template<typename LengthUnit, typename AngleUnit>
+class Pose2
+  : public std::tuple<Vector2<LengthUnit>, AngleUnit>
+{
+    static_assert(units::traits::is_length_unit<LengthUnit>::value,
+                  "LengthUnit is not a unit of length");
+    static_assert(units::traits::is_angle_unit<AngleUnit>::value,
+                  "AngleUnit is not a unit of angle");
+
+public:
+    Pose2() = default;
+
+    Pose2(LengthUnit x, LengthUnit y, AngleUnit angle)
+      : std::tuple<Vector2<LengthUnit>, AngleUnit>({ x, y }, angle)
+    {}
+
+    template<typename LengthUnit2, typename AngleUnit2>
+    operator Pose2<LengthUnit2, AngleUnit2>() const
+    {
+        return Pose2<LengthUnit2, AngleUnit2>{ x(), y(), yaw() };
+    }
+
+    template<typename LengthUnit2, typename AngleUnit2>
+    operator Pose3<LengthUnit2, AngleUnit2>() const
+    {
+        return Pose3<LengthUnit2, AngleUnit2>{ { x(), y(), z() }, { yaw(), pitch(), roll() } };
+    }
+
+    Vector2<LengthUnit> &position() { return std::get<0>(*this); }
+    const Vector2<LengthUnit> &position() const { return std::get<0>(*this); }
+    LengthUnit &x() { return std::get<0>(*this)[0]; }
+    const LengthUnit &x() const { return std::get<0>(*this)[0]; }
+    LengthUnit &y() { return std::get<0>(*this)[1]; }
+    const LengthUnit &y() const { return std::get<0>(*this)[1]; }
+    static constexpr LengthUnit z() { return LengthUnit(0); }
+
+    Array3<AngleUnit> attitude() const { return { yaw(), AngleUnit(0), AngleUnit(0) }; }
+    AngleUnit &yaw() { return std::get<1>(*this); }
+    const AngleUnit &yaw() const { return std::get<1>(*this); }
+    static constexpr AngleUnit pitch() { return AngleUnit(0); }
+    static constexpr AngleUnit roll() { return AngleUnit(0); }
```

Comment:
Again I don't really like this way of making different types of pose interchangable. How about just doing away with ``Pose2`` and use ``Pose3`` for everything? Those extra bytes almost certainly don't matter and 2D stuff can just leave alone the components they don't use (also quite a lot of 2Dish stuff like the antworld are essentially 2D but z isn't at zero so they hover above the ground a bit...

Reply:
We could use ``Pose3`` in most places, but there are places where we are only saving x, y and angle where ``Pose2`` would fit better. Performance-wise it won't matter if we have one ``Pose3`` object, but it might do for e.g. a really big array of Vicon data and if we don't care about the other dimensions it doesn't seem like there's much point in saving them... I'm not totally against the idea, but I feel like it makes sense to have a separate class for this.

I agree it's a bit weird to have e.g. a ``z()`` for a ``Pose2``, but it's mostly for some of the simulation bits and pieces I've done on other branches: the idea was that you can have template functions which take a ``PoseType`` as an argument and then just write a generic function using ``x()``, ``y()`` and ``z()`` etc.

I've used some variation on this code for the robot positioner stuff and for other simulation bits I've been doing, but I thought I'd do it as a separate PR because it touches a lot of code and I thought it might be the most controversial bit ;-). I don't mind hacking away away at it some more but it would be great if we could get it merged in some form soon-ish just so I can rebase the new robot positioner/visual navigation bits off it :-). Obvs I don't want you to merge anything that's too obviously nasty though!